### PR TITLE
Updated fix for #470 (tx signing issues).  Now using kyuupichan's advice to pass the sequence number from the wallet down

### PR DESCRIPTION
**Pull request by cculianu**
Tested, works.  Has 0 side-effects other than fixing #470.  I recommend merging.

Taken from a fix proposed here: 

https://github.com/goblin/electrum/commit/a3db80c64831b46e255c0fc530f14322530a2b2e

### CODE REVIEW

**Comment by kyuupichan** [comment]
Why do we need to scan?  Why can the wallet not find the address in its address list, and go from there?

**Comment by cculianu** [comment]
Because public key type 'fd' doesn't include the sequence number.

The deterministic wallet doesn't always keep public/private key pairs around -- it computes them based on the sequence number.

'ff' public key type from the tx encodes this ID number.

'fd' (the support we are adding) doesn't.  Have to compute it.

I can't figure out another way to do it.  If you can -- be my guest.

**Comment by cculianu** [comment]
@kyuupichan  Hey man!  Ok, I took your advice and passed the information into the keystore from the wallet using a new 'advisory' (optional) mechanism called "wallet_advice".  

I tested this with all combinations of wallets and the codepath is ONLY really taken on Bip32 wallets with type 'fd' tx's that they must sign (as in the situation in #470).  

For everything else this code is harmless/does nothing.

It's much faster and can handle anything.

I'm sorta glad you made me do it this way.  

Looking forward to closing this issue, so let me know if anything else is wrong/right with this.  Thanks. 👍

File: lib/keystore.py [hunk 1]
```diff
@@ -73,6 +77,9 @@ def can_sign(self, tx):
             return False
         return bool(self.get_tx_derivations(tx))
 
+    def set_wallet_atvice(self, addr, advice):
```
Comment (SomberNight) [inline lib/keystore.py]: typo `wallet_advice `

**Comment by cculianu** [comment]
Tee-hee.. sorry! Fixed the typo. 👍

**Comment by kyuupichan** [comment]
I'm really not happy with this implementation.  Unfortunately @fyookball committed it.  I'll try and rework it.

**Comment by cculianu** [comment]
Ha! Go for it my man.  They say the best way to get an answer to something on the internet is to post a question or a comment with a half-wrong explanation.

Similarly, one of the best ways to get people to contribute to open source is to post a less-than-ideal implementation.

I, too, feel it could be cleaned up from an OO/encapsulation point of view.

Go for it man!  Glad I could incite you into action.  ;P

**Comment by fyookball** [comment]
@kyuupichan your advice is always valued, but just curious what you didnt like about this philosophically... it makes a lot of sense to me.

**Comment by kyuupichan** [comment]
A lot of things - shouldn't be storing state in the keystore, should not be fixed just for BIP32 wallets, doesn't work for hardware signing, ....

**Comment by cculianu** [comment]
> should not be fixed just for BIP32 wallets

AFAICT the issue only affects BIP32 wallets.

As for the issue with "should not be storing state in keystore" -- that's just your opinion.

**Pull request merged**